### ISSUE TITLE
[Rancher 2.0] cluster owner hasn't enough permissions

### ISSUE THREAD

**Issue by k0nstantinv** [rancher/rancher#13058]
My infrastructure:
Ubuntu 16.04 xenial servers ( kernel: 4.4.0-119-generic )
```
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.4 LTS
Release:	16.04
Codename:	xenial
```
**docker:** `17.03.2-ce`
**rancher-server**: `rancher/server:preview` `v2.0.0-beta3`
**rancher-agent:** `rancher/agent:v2.0.0-beta3`

**kubernetes version:** `v1.9.5-rancher1-1`
**network provider**: `flannel`
**ingress provider**: `Nginx`
everything else by default

I have a single node Rancher server with SSL in docker container. 

offtopic: _I'm trying to integrate k8s cluster with the GitLab. And everything looks very simple if you are using native Kubernetes. But In case of cluster, which was created by the Rancher misunderstanding appears. Generally, this requires Cluster Name, API URL for integration. First step - install Helm (and Tiller) to cluster from newly integrated Gitlab. Gitlab, in turn, will try to create a namespace `gitlab-managed-apps` inside of the k8s custer_

Issue itself:

1. I have a Rancher user, he has administrator permissions for everything and he is my k8s cluster owner as well
2. I use token for this user from his `kubeconfig` file, something like: "kubeconfig-user-xxxxx:cczbq98n5vwmvlscxm7k8zkqj9xkrb7qdqvvk8m8g6x94zcpk7wjbp"
3. I use API URL: "https://<my_server>/k8s/clusters/cluster-xxxxx"
4. I see errors from the Gitlab side: 
```
Kubernetes error: Namespace in version "v1" cannot be handled as a Namespace: no kind "Namespace" is registered for version "clusters/v1"
```
5. As a workaround i've created the namespace `gitlab-managed-apps` manually
6. Then i see another error: 
```
Kubernetes error: Pod in version "v1" cannot be handled as a Pod: no kind "Pod" is registered for version "clusters/v1"
```
So, It can be assumed that my user still has not enough rights for creating k8s resources

**Comment by k0nstantinv** [comment]
> Judging by the pattern there seems to be some auto-magic going on inside Gitlab using the provided k8s URL which gets tripped up by Rancher's prefixed path components.

Oh thanks! I think it is... and i can ask GitLab how they parse API URL for integration.
> I worked around the problem by providing a reverse proxy on a separate vhost which strips Ranchers URL prefixing and provides access to the k8s api-server anchored at / (for every API endpoint provided by k8s and passing every other URL to Rancher as is to keep auth etc. intact).
did it work?

**Comment by dirkd** [comment]
It does indeed work.
I even stripped down the reverse proxy to only provide the k8s api (so no Rancher endpoints like /login) since Gitlab's integration won't use them anyway and the proxy is only accessible by Gitlab in our case. I'm currently corresponding with Gitlab's enterprise support to maybe resolve this issue on their end so that Rancher's default api-server URLs can be used as is without any of this fiddling.

I've uploaded a masked version of our nginx config (derived from Rancher 1.6's default reverse proxy config in the docs, variables in double braces and allcaps) at https://gist.github.com/dirkd/b0f2de437da5e602f00083ebc5c3a467 for your convenience. While Rancher 2 provides direct access on :80 and :443 on the rancher/server container the config should integrate fine on a separate ingress endpoint (another host, IP address, port or any combination of those).

**Comment by dirkd** [comment]
addendum: The problem seems to be the kubeclient gem used by Gitlab. Initializing it with a Rancher API-Server URL yields a client which has its `@api_group` attribute set to the wrong value:

```ruby
2.4.1 :001 > require 'kubeclient'
 => true
2.4.1 :002 > ::Kubeclient::Client.new("https://testhost/r/projects/1a7/kubernetes:6443/", "v1", auth_options: {}, ssl_options: { verify_ssl: OpenSSL::SSL::VERIFY_PEER }, http_proxy_uri: "")
 => #<Kubeclient::Client:0x00000001eadb60 @api_endpoint=#<URI::HTTPS https://testhost/r/projects/1a7/kubernetes:6443>, @api_group="projects/", @entities={}, @discovered=false, @api_version="v1", @headers={}, @ssl_options={:verify_ssl=>1}, @auth_options={}, @socket_options={:socket_class=>nil, :ssl_socket_class=>nil}, @timeouts={:open=>60, :read=>60}, @http_proxy_uri="">
2.4.1 :003 > ::Kubeclient::Client.new("https://testhost/k8s/clusters/cluster-1a2b3", "v1", auth_options: {}, ssl_options: { verify_ssl: OpenSSL::SSL::VERIFY_PEER }, http_proxy_uri: "")
 => #<Kubeclient::Client:0x000000019ec4a0 @api_endpoint=#<URI::HTTPS https://testhost/k8s/clusters/cluster-1a2b3>, @api_group="clusters/", @entities={}, @discovered=false, @api_version="v1", @headers={}, @ssl_options={:verify_ssl=>1}, @auth_options={}, @socket_options={:socket_class=>nil, :ssl_socket_class=>nil}, @timeouts={:open=>60, :read=>60}, @http_proxy_uri="">
```

So this issue is definitely not related to Rancher but https://github.com/abonas/kubeclient/tree/v3.0.0 and should probably be closed here.

**Comment by k0nstantinv** [comment]
@dirkd great research! This issue undoubtedly should be closed

**Comment by dmastag** [comment]
@dirkd I already have my Rancher 2.0 working with HA using an external Load Balancer to access the GUI

But I can't figure out how to use your gist so that gitlab can connect to it. Could you provide some pointers ?

**Comment by dmastag** [comment]
I can access the api using

https://rancher.server.com/k8s/clusters/local/

Which nicely brings up the API that I assume I must proxy pass

![image](https://user-images.githubusercontent.com/5745805/41704078-7568d920-755f-11e8-98c8-cb7ded0523f6.png)

Then I try to create a new subdomain that does a proxy pass into that API

```
  server {
      listen 443 ssl http2;
      server_name rancherapi.server.com;
      ssl_certificate "";
      ssl_certificate_key "";

      location / {
          # rewrite ^/k8s/clusters/local/(.*)$ /$1 break;
          proxy_pass https://rancher.server.com/k8s/clusters/local/;
          proxy_http_version 1.1;
          proxy_set_header Upgrade $http_upgrade;
          proxy_set_header Connection 'upgrade';
          proxy_set_header Host $host;
          proxy_cache_bypass $http_upgrade;
      }

  }
```

But for some reason I am getting 414 Request-URI Too Large

![image](https://user-images.githubusercontent.com/5745805/41704184-c81f983e-755f-11e8-868b-4990137dc433.png)

And can see that the url is forever adding to itself making the url go to

https://rancher.server.com/k8s/clusters/localk8s/clusters/localk8s/clusters/localk8s/clusters/localk8s/clusters/localk8s/clusters/localk8s/clusters/localk8s/clusters/localk8s/clusters/localk8s/clusters/localk8s/clusters/local

Am I getting closer @k0nstantinv ?

**Comment by dirkd** [comment]
You need to set the proxied Host header to the backend's hostname (i.e. `rancher.server.com`) as indicated by my example configuration.

By using `proxy_set_header Host $host` you send the frontend's hostname (`rancherapi.server.com`) to the backend which then tries to serve your request by redirecting to the domain it knows about which creates the loop.
Since you're accessing the backend by hostname in your proxy_pass instead of just an IP address you can also omit setting the `proxy_set_header Host` and let nginx take care of sending the correct Host header.

TL;DR: `s/proxy_set_header Host $host;/proxy_set_header Host rancher.server.com;/`

**Comment by k0nstantinv** [comment]
@dmastag, unfortunately i have no idea about that, i'm not working with Rancher at this time. But i think that passing all this steps just to make GitLab could parse Rancher's k8s API URL is quite unnormal, isn't it?

**Comment by dmastag** [comment]
@dirkd Thanks
Though I am sadly still getting an error like below

![image](https://user-images.githubusercontent.com/5745805/41840676-876e9400-7890-11e8-90eb-b2877ec5ec76.png)

**Comment by marconett** [comment]
fyi: there's no need for a nginx proxy if your gitlab can access the kubernetes api on the rancher host directly via `https://rancher:6443`. just enter that domain as `API URL` in gitlab.

**Comment by dmastag** [comment]
Thanks, so I finally got it working by combining a couple of Articles

I used the Internal IP as @marconett, but since its https I had to extract the PEM certificate.

So I based on this Answer on the gitlab forum got the PEM
https://forum.gitlab.com/t/kubernetes-integration-something-went-wrong-while-installing-helm-tiller/15627/6

I had a problem with the Token and also got the following problem
"Kubernetes error: namespaces "gitlab-managed-apps" is forbidden: User "system:serviceaccount:gitlab-managed-apps:gitlab-sa" cannot get namespaces in the namespace "gitlab-managed-apps"

Luckily I was able to find the following article.

http://centosquestions.com/kubernetes-error-namespaces-gitlab-managed-apps-forbidden-user-systemserviceaccountgitlab-managed-appsgitlab-sa-cannot-get-namespaces-namespace-gitlab-managed-apps/

**Comment by ru-it-specialist** [comment]
For:

> Kubernetes error: namespaces "gitlab-managed-apps" is forbidden: User "system:serviceaccount:gitlab-managed-apps:gitlab-sa" cannot get namespaces in the namespace "gitlab-managed-apps`

try:

```sh
kubectl create clusterrolebinding --user system:serviceaccount:gitlab-managed-apps:default default-gitlab-sa-admin --clusterrole cluster-admin
```

I don't think that disabling RBAC is good idea.

**Comment by gabrielrcouto** [comment]
I had the same "Namespace in version "v1" cannot be handled as a Namespace...." problem with my cluster on AWS. To solve, I  created an ALB with HTTPS and a valid certificate for my domain (mycluster.mydomain.com) and pointed to Kubernetes ControlPlane instances on port 6443 (on the health check, use status code 200-499). On Gitlab, I used "https://mycluster.mydomain.com" and not filled the CA certificate field.

**Comment by Go-Macyun** [comment]
> addendum: The problem seems to be the kubeclient gem used by Gitlab. Initializing it with a Rancher API-Server URL yields a client which has its `@api_group` attribute set to the wrong value:
> 
> ```ruby
> 2.4.1 :001 > require 'kubeclient'
>  => true
> 2.4.1 :002 > ::Kubeclient::Client.new("https://testhost/r/projects/1a7/kubernetes:6443/", "v1", auth_options: {}, ssl_options: { verify_ssl: OpenSSL::SSL::VERIFY_PEER }, http_proxy_uri: "")
>  => #<Kubeclient::Client:0x00000001eadb60 @api_endpoint=#<URI::HTTPS https://testhost/r/projects/1a7/kubernetes:6443>, @api_group="projects/", @entities={}, @discovered=false, @api_version="v1", @headers={}, @ssl_options={:verify_ssl=>1}, @auth_options={}, @socket_options={:socket_class=>nil, :ssl_socket_class=>nil}, @timeouts={:open=>60, :read=>60}, @http_proxy_uri="">
> 2.4.1 :003 > ::Kubeclient::Client.new("https://testhost/k8s/clusters/cluster-1a2b3", "v1", auth_options: {}, ssl_options: { verify_ssl: OpenSSL::SSL::VERIFY_PEER }, http_proxy_uri: "")
>  => #<Kubeclient::Client:0x000000019ec4a0 @api_endpoint=#<URI::HTTPS https://testhost/k8s/clusters/cluster-1a2b3>, @api_group="clusters/", @entities={}, @discovered=false, @api_version="v1", @headers={}, @ssl_options={:verify_ssl=>1}, @auth_options={}, @socket_options={:socket_class=>nil, :ssl_socket_class=>nil}, @timeouts={:open=>60, :read=>60}, @http_proxy_uri="">
> ```
> 
> So this issue is definitely not related to Rancher but https://github.com/abonas/kubeclient/tree/v3.0.0 and should probably be closed here.

Hello, I used the abonas/kubeclient gem and met the same question.Could you help me to solve this?
Here is my issue: https://github.com/abonas/kubeclient/issues/418?tdsourcetag=s_pcqq_aiomsg

**Comment by codejamninja** [comment]
I created a docker container and a rancher compatible helm chart for proxying the rancher kubernetes api so gitlab can understand it.

https://github.com/codejamninja/kube-api-proxy
